File: python/src/sum_count.py
from common import time_execution


@time_execution()
def count_sum_dp(target_sum: int, numbers: list) -> int:
    dp = [0] * (target_sum + 1)
    dp[0] = 1  # There is one way to make zero (with no numbers)
    for num in numbers:
        # To get from 7 to 10, we can for example add 3
        # So for all ways that we can construct 7, we can also add 3 and get to 10
        for n in range(num, target_sum + 1):
            dp[n] += dp[n - num]

    return dp[-1]


@time_execution()
def count_sum(target_sum: int, numbers: list[int]) -> int:
    def helper(idx: int, sum_so_far: int = 0) -> int:
        if sum_so_far == target_sum:
            return 1
        if idx == len(numbers) or sum_so_far > target_sum:
            return 0

        return helper(idx, sum_so_far+numbers[idx]) + helper(idx+1, sum_so_far)

    return helper(0)


@time_execution()
def count_sum_memo(target_sum: int, numbers: list[int]) -> int:
    memo = {target_sum: 1}
    memo_count = 0

    def helper(idx: int, sum_so_far: int = 0) -> int:
        nonlocal memo_count
        key = (sum_so_far, idx)

        if sum_so_far == target_sum:
            return 1
        if key in memo:
            memo_count += 1
            return memo[key]
        if idx == len(numbers) or sum_so_far > target_sum:
            return 0

        memo[key] = helper(
            idx, sum_so_far+numbers[idx]) + helper(idx+1, sum_so_far)
        return memo[key]

    return f"{helper(0)}, {memo_count} memo lookups"


@time_execution()
def count_sum_v2(target_sum: int, numbers: list) -> int:
    def helper(sum_so_far: int, last_used_index: int) -> int:
        if sum_so_far == target_sum:
            return 1
        if sum_so_far > target_sum:
            return 0

        count = 0
        # Only consider numbers at or after the last used index to prevent permutations of the same combination
        for i in range(last_used_index, len(numbers)):
            count += helper(sum_so_far + numbers[i], i)

        return count

    return helper(0, 0)


@time_execution()
def count_sum_v2_memo(target_sum: int, numbers: list) -> int:
    memo = {}
    memo_count = 0

    def helper(sum_so_far: int, last_used_index: int) -> int:
        nonlocal memo_count
        key = (sum_so_far, last_used_index)

        if sum_so_far == target_sum:
            return 1
        if key in memo:
            memo_count += 1
            return memo[key]
        if sum_so_far > target_sum:
            return 0

        count = 0
        # Only consider numbers at or after the last used index to prevent permutations of the same combination
        for i in range(last_used_index, len(numbers)):
            count += helper(sum_so_far + numbers[i], i)

        memo[key] = count
        return count

    return f"{helper(0, 0)}, {memo_count} memo lookups"


print(count_sum_dp(35, [1, 2, 3]))
print(count_sum(35, [1, 2, 3]))
print(count_sum_memo(35, [1, 2, 3]))
print(count_sum_v2(35, [1, 2, 3]))
print(count_sum_v2_memo(35, [1, 2, 3]))

File: php/src/SumCount.php
<?php

function sum_count_dp(int $target, array $arr): int
{
    $dp = array_fill(0, $target + 1, 0);
    $dp[0] = 1;
    foreach ($arr as $num) {
        // Increment ways to get to 'i' by ways to get to i-num, because we can add num and get to 'i'
        for ($i = $num; $i <= $target; $i++)
            $dp[$i] += $dp[$i - $num];
    }

    return $dp[$target];
}

function sum_count_rec(int $target, array $arr): int
{
    $helper = function (int $sum_so_far, int $last_used_idx = 0) use (&$helper, $arr): int {
        if ($sum_so_far < 0)
            return 0;
        if ($sum_so_far == 0)
            return 1;

        $count = 0;
        // Start from last used so we don't get permutations
        for ($i = $last_used_idx; $i < count($arr); $i++)
            $count += $helper($sum_so_far - $arr[$i], $i);

        return $count;
    };
    return $helper($target);
}

function sum_count_memo(int $target, array $arr): int
{
    $memo = [];
    $memo_hits = 0;
    $helper = function (int $sum_so_far, int $last_used_idx = 0) use (&$helper, &$memo, &$memo_hits, $arr): int {
        $key = serialize([$sum_so_far, $last_used_idx]);
        if (isset($memo[$key])) {
            $memo_hits++;
            return $memo[$key];
        }
        if ($sum_so_far < 0)
            return 0;
        if ($sum_so_far == 0)
            return 1;

        $count = 0;
        // Start from last used so we don't get permutations
        for ($i = $last_used_idx; $i < count($arr); $i++)
            $count += $helper($sum_so_far - $arr[$i], $i);

        $memo[$key] = $count;
        return $count;
    };
    $res = $helper($target);
    //echo "Memo hits:$memo_hits", "\n";
    return $res;
}

function sum_count_rec_v2(int $target, array $arr): int
{
    $helper = function (int $sum_so_far, int $idx = 0) use (&$helper, $arr): int {
        if ($sum_so_far < 0 || $idx == count($arr))
            return 0;
        if ($sum_so_far == 0)
            return 1;

        return $helper($sum_so_far - $arr[$idx], $idx) + $helper($sum_so_far, $idx + 1);
    };
    return $helper($target);
}


function sum_count_memo_v2(int $target, array $arr): int
{
    $memo = [];
    $memo_hits = 0;
    $helper = function (int $sum_so_far, int $idx = 0) use (&$helper, &$memo, &$memo_hits, $arr): int {
        $key = serialize([$sum_so_far, $idx]);
        if (isset($memo[$key])) {
            $memo_hits++;
            return $memo[$key];
        }
        if ($sum_so_far < 0 || $idx == count($arr))
            return 0;
        if ($sum_so_far == 0)
            return 1;

        $memo[$key] = $helper($sum_so_far - $arr[$idx], $idx) + $helper($sum_so_far, $idx + 1);
        return $memo[$key];
    };

    $res = $helper($target);
    //echo "Memo hits:$memo_hits", "\n";
    return $res;
}